compiler/noirc_evaluator/src/ssa/checks/check_for_underconstrained_values.rs

```diff
@@ -63,6 +92,355 @@ fn check_for_underconstrained_values_within_function(
+    Unconstrained { descendants: HashSet<ValueId> },
+    Constrained,
+}
+
+impl BrilligTaintedIds {
+    fn new(arguments: &[ValueId], results: &[ValueId]) -> Self {
+        BrilligTaintedIds {
+            arguments: HashSet::from_iter(arguments.iter().copied()),
+            results: results
+                .iter()
+                .map(|result| ResultStatus::Unconstrained { descendants: HashSet::from([*result]) })
+                .collect(),
+            root_results: HashSet::from_iter(results.iter().copied()),
+        }
+    }
+
+    /// Add children of a given parent to the tainted value set
+    /// (for arguments one set is enough, for results we keep them
+    /// separate as the forthcoming check considers the call covered
+    /// if all the results were properly covered)
+    fn update_children(&mut self, parent: &ValueId, children: &[ValueId]) {
+        if self.arguments.contains(parent) {
+            self.arguments.extend(children);
+        }
+        for result_status in &mut self.results.iter_mut() {
+            match result_status {
+                // Skip updating results already found covered
+                ResultStatus::Constrained => {
+                    continue;
+                }
+                ResultStatus::Unconstrained { descendants } => {
+                    if descendants.contains(parent) {
+                        descendants.extend(children);
+                    }
+                }
+            }
+        }
+    }
+
+    /// If Brillig call is properly constrained by the given ids, return true
+    fn check_constrained(&self) -> bool {
+        // If every result has now been constrained,
+        // consider the call properly constrained
+        self.results.iter().all(|result| matches!(result, ResultStatus::Constrained))
+    }
+
+    /// Remember partial constraints (involving some of the results and an argument)
+    /// along the way to take them into final consideration
+    fn store_partial_constraints(&mut self, constrained_values: &HashSet<ValueId>) {
+        let mut results_involved: Vec<usize> = vec![];
+
+        // For a valid partial constraint, a value descending from
+        // one of the results should be constrained
+        for (i, result_status) in self.results.iter().enumerate() {
+            match result_status {
+                // Skip checking already covered results
+                ResultStatus::Constrained => {
+                    continue;
+                }
+                ResultStatus::Unconstrained { descendants } => {
+                    if descendants.intersection(constrained_values).next().is_some() {
+                        results_involved.push(i);
+                    }
+                }
+            }
+        }
+
+        // Also, one of the argument descendants should be constrained
+        // (skipped if there were no arguments, or if an actual result and not a
+        // descendant has been constrained, e.g. against a constant)
```

Comment:
Can you add a comment of why this is the rule?

Reply:
Cleared the relationship between the result and argument descendant up a bit